### PR TITLE
changed SELECT mode default to "set" and added FileUpdater that migrates existing buttons

### PR DESCRIPTION
**Pull request by ArnoldSmith86**
This is the first change that would break existing games.

I don't think it's the most important change in the world but it's probably the most irritating default at the moment.

The main purpose of this pull request was to build a method to introduce breaking changes so that we don't limit ourselves to the status quo.

Please do a few tests if this is working as intended (no behavior of any existing game is allowed to be changed by this).

### CODE REVIEW

**Comment by ArnoldSmith86** [comment]
PR-SERVER-BOT: You can play around with it here: http://212.47.248.129:3083/pr-test (or any other room on that server)

**Comment by 96LawDawg** [comment]
I love the idea of this change.  But it is a little broke.  First thing I did was run the SELECT tutorial.  Hit the default button (which specified add mode) and it breaks.

**Comment by 96LawDawg** [comment]
Correction.  I just looked at the original tutorial and the code went **_from this:_**

{
      "func": "SELECT"
    },
    {
      "func": "SET",
      "property": "css",
      "value": null
    },
    {
      "func": "SELECT",
      "property": "id",
      "value": "8fr3",
      "mode": "set"
    },
    {
      "func": "SET",
      "property": "css",
      "value": "font-size: 60px"
    }

**_to this_** 

{
      "func": "SELECT",
      "mode": "add"
    },
    {
      "func": "SET",
      "property": "css",
      "value": null
    },
    {
      "func": "SELECT",
      "property": "id",
      "value": "8fr3",
      "mode": "add"
    },
    {
      "func": "SET",
      "property": "css",
      "value": "font-size: 60px"
    }

**Comment by RaphaelAlvez** [comment]
Would it be better to check all presets we would like to change and make just one or with all of them?

**Comment by UltimateGeek** [comment]
Some considerations on managing the breaking change:
1. Establish an integration branch for multiple PRs to promote to before merging to to master.
2. Collect multiple breaking changes together before promoting
3. Once ready, establish a timeline to promote the breaking changes. Immediately update master with a notice on the production site and the wiki to indicate the date of it going live
4. Have an integration branch also in the library repo for iterating on updates for the breaking changes
5. Change the library’s src to be flat files instead of zip files and establish a build. This will make it much easier to make updates in parallel before making the transition

**Comment by pf-lkobbe** [comment]
I like the idea of collecting breaking changes. First of all, it reduces chances that people are missing an important change. Second, it gives a better opportunity for managing updates tutorials and documentation, for instance by splitting up the subsequent tasks.

**Comment by ArnoldSmith86** [comment]
I'm not sure if you understand this correctly but the point would be to **not** break anything. Apparently that didn't work yet; I look into that.

But I don't see why this would need such a big process.

The only change really would be that game designers need to know the new default for new buttons they build.

**Comment by ArnoldSmith86** [comment]
The failing check is pretty impressive and definitely an issue (I'm aware of) but why does it fail here and now? Did you change something, @UltimateGeek?

**Comment by UltimateGeek** [comment]
> The failing check is pretty impressive and definitely an issue (I'm aware of) but why does it only fail here and now? Did you change something, @UltimateGeek?

No, this is to help prevent another security issue from being created by your code change.

Please sanitize user inputs. We don’t want the user to cause the server to overwrite files on the server if they had the room id be something like “../../../etc/passwd”

**Comment by ArnoldSmith86** [comment]
I'm pretty sure I didn't introduce a new issue. But I guess it gets flagged because I changed the line that has the issue.

File: server/fileupdater.mjs
Comment (robartsd) [inline server/fileupdater.mjs:13]: Would there be a problem if .clickRoutine is undefined?

File: server/fileupdater.mjs
Comment (robartsd) [inline server/fileupdater.mjs:19]: I'm not clear on what isFirstSelect is accomplishing here? Looks to me like for each clickRoutine it prevents updating subsequent 'SELECT' functions without mode set to use mode = 'add'. Won't this cause incorrect conversions?

File: server/fileupdater.mjs
Comment (ArnoldSmith86) [reply server/fileupdater.mjs:13]: Yes. I'll fix that.

File: server/fileupdater.mjs
Comment (ArnoldSmith86) [reply server/fileupdater.mjs:19]: The idea is: if it's the first `SELECT` in the button, the mode doesn't make a difference. That's why I usually don't set the `mode` parameter there when I make a new button. I feel like it's weird to make the conversion add `mode: add` to those though.

Comment (rogerl50) [reply server/fileupdater.mjs:19]: Is it your intention to leave `fileupdater` in the code for the indefinite future? If you hope to use it for any other incompatible changes that might get made moving forward, then the `VERSION` hack you use to check if the game in question has already been converted doesn't seem like it will work the next time. You might want to make it explicit to `SELECT` by renaming it. OTOH if you expect this file to apply only to this issue, perhaps the file itself should be renamed.

Comment (rogerl50) [reply server/fileupdater.mjs:19]: If I understand this correctly, `fileupdater` will run each time a room is loaded, even though the changes it makes will be persisted if the room is saved to disk. Is that correct? If so, is it possible to put a bit of metadata in the room's JSON that contains this bit of version information? That would reduce load time (though obviously not by much), and might come in handy if there are other future changes that require a similar update change. (And, if this happens, the `VERSION` tag should be changed to something more specific to this particular issue.)

Comment (ArnoldSmith86) [reply server/fileupdater.mjs:19]: It does. Every state has a `state._meta.version`. The FileUpdater checks if it is already equal to `VERSION` (the latest version of the state format that is expected by the code) and if it is, it doesn't do anything.

With the next change I would increment `VERSION` to `3` and check with `v < 3` if the change has to be applied to the state.

Comment (rogerl50) [reply server/fileupdater.mjs:19]: That makes sense. I didn't see anything named `state` or `version` in a CURL dump of a room's `JSON`, which made me think that it wasn't persisted. Where is it?

Comment (ArnoldSmith86) [reply server/fileupdater.mjs:19]: Yes, that API removes the version information and is not meant for long term storage somewhere. Might be better to add it. In VTT files you can see the `_meta`. For the room itself the `_meta` also includes the list of games and players etc.

Comment (robartsd) [reply server/fileupdater.mjs:19]: OK, that makes sense to me now.

**Review by rogerl50 — APPROVED** [review]

**Review by 96LawDawg — APPROVED** [review]
Tested by checking on multiple tutorials and games.  Everything seems to work fine now.  I'm going be the second approver, but will not merge in case ArnoldSmith86 knows of something else.

**Comment by ArnoldSmith86** [comment]
The plan is to wait for #147, I think. Have these two as the first two changes that require the FileUpdater and then collect the next ones for four weeks.

**Comment by robartsd** [comment]
> The plan is to wait for #147, I think. Have these two as the first two changes that require the FileUpdater and then collect the next ones for four weeks.

#147 certainly is worth holding other FileUpdater changes for.

**Comment by larkob** [comment]
I just looked at the `FileUpdater` and noticed that only the first SELECT would get mode=set whereas any subsequent SELECTs with missing mode would get mode=add. That's not what I had expected from reading the pull request description. I thought every SELECT without specific mode would get convert to mode=set. Ultimately, we should do whatever we believe to cause less problems in the migration, but it's probably worth pointing out.

**Comment by robartsd** [comment]
> I just looked at the `FileUpdater` and noticed that only the first SELECT would get mode=set whereas any subsequent SELECTs with missing mode would get mode=add. That's not what I had expected from reading the pull request description. I thought every SELECT without specific mode would get convert to mode=set. Ultimately, we should do whatever we believe to cause less problems in the migration, but it's probably worth pointing out.

I thought this when I first looked at the code too, but it is the opposite. The first SELECT is skipped for adding .mode="add" where .mode was missing since any collection would start empty.

**Comment by larkob** [comment]
Ah, I see. The PR title really threw me off. How come the routine deletes any existing mode=get?

**Comment by rogerl50** [comment]
> Ah, I see. The PR title really threw me off. How come the routine deletes any existing mode=set?

I assumed that it was just because it is now the default, so all having it there does is increase the amount of network traffic.

**Comment by ArnoldSmith86** [comment]
I mean, there's no way to know if the game designer wants the default stated there explicitly or not. TBH I only thought about my preference which is to not add any defaults.

Do you guys think they should be left in?

**Comment by robartsd** [comment]
When I think the defaults are good, I don't mind leaving out the property. Since I dislike "mode" defaulting to "add", I'm in the habit of using "mode" explicitly all the time. I think I will change my habit after this pull request is merged.

**Comment by larkob** [comment]
The code only changes the first SELECT in any clickRoutine. What if the first SELECT applies to a different collection than the second SELECT in a clickRoutine? Shouldn't the principle apply to the first SELECT per collection?

```
{
  func: SELECT,
  ...
  collection: collectionA
},
{
  func: SELECT
  ...
  collection: collectionB
}

becomes in v2:

{
  func: SELECT,
  ...
  collection: collectionA
},
{
  func: SELECT
  ...
  collection: collectionB
  mode: add
}
```

**Comment by ArnoldSmith86** [comment]
True. I didn't think that gets used all that much at the time but that probably changed...

**Comment by RaphaelAlvez** [comment]
That can totally be done in another PR since it would definitely be the coder mistake

**Comment by ArnoldSmith86** [comment]
It can't and it isn't. The point is that for the first `SELECT` to any collection, it doesn't matter if the `mode` is `add` or `set`; it does the same. So the file updater shouldn't add it for those.

**Comment by ArnoldSmith86** [comment]
Loading v2 files doesn't work. There's a `variant._meta.version !== 1` check in there but I gotta do that differently now so that the current version number only has to be changed in one place.

**Pull request merged**
